In [42]:
import numpy as np
import pandas as pd
import statsmodels.api as sm

---

## Økonometri A

# Problem set 5

---



### Vækstregressioner: Absolut og betinget konvergens

I denne ugeseddel skal I udvide regressionsmodellen fra ugeseddel 3, hvor I så på hypotesen om absolut og betinget konvergens, ved at inkludere information om hvert lands humankapital. I Makroøkonomi A vil I se, at en model uden humankapital vil underestimere betydningen af opsparing og befolkningsvækst på BNP pr. arbejder og samtidig overestimere hastigheden af, hvor hurtigt et land konvergerer mod en vækstbane.

I skal tilføje en variabel i regressionsmodellen, som viser investeringsraten i humankapital korrigeret for befolkningsudviklingen, afskrivningsrate og teknologisk fremgang. I skal ligesom i ugeseddel 3 anvende en afskrivningsrate og vækst i teknologi på 0,075. Som en udvidelse til ugeseddel 3 skal I formelt teste forskellige hypoteser om absolut og betinget konvergens. I skal forklare, hvornår I skal bruge et _F_-test og et _t_-test og under hvilke antagelser, de giver valide resultater.

Regressionsmodellen med humankapital kan omskrives til:

$$
gy_i = \beta_0 + \beta_1 \log (y60_i) + \beta_2 strucK_i + \beta_3 strucH_i + \epsilon_i \tag{1}
$$

hvor de strukturelle karakteristika er 
- $ strucK_i = \ln(sk_i) - \ln(n_i + 0,075) $ 
- og $  strucH_i = \ln(sh_i) - \ln(n_i + 0,075) $

I skal anvende det samme datasæt, som I brugte til ugeseddel 3.



---

### Gruppespørgsmål


#### Opgave 1 
I kan pålægge en model parameterrestriktioner for at omskrive den til en mere simpel model og indføre disse som en nulhypotese. Den udvidede Solow-model (1) indeholder også den simple regressionsmodel fra spørgsmål 4 i ugeseddel 3, hvor I så på hypotesen om absolut konvergens. Hvad er parameterrestriktionerne og derved nulhypotesen i den udvidede model (1), hvis I gerne vil teste hypotesen om absolut konvergens? Hvad er alternativ hypotesen?



**Dit svar:**
$H_0$: $\beta_2=\beta_3=0$ altså IKKE betinget konvergens. Alternativ er at blot en af dem er forskellig fra 0.

#### Opgave 2 
Giv et forslag til et relevant test, som I kan bruge til at teste nulhypotesen i gruppespørgsmål 1. Hvad er antallet af parameterrestriktioner, som I gerne vil teste?



**Dit svar:**

F test med 2 parameterrestriktioner

#### Opgave 3

Hvilke antagelser skal være opfyldt for, at det foreslåede test i gruppespørgsmål 2 er validt?



**Dit svar:**

MLR 1-6

---

### Python øvelser



#### Opgave 1
Indlæs data i Python. Fjern de lande, som har datakvalitet D, og lande, som mangler observationer for investeringsraten i humankapital, **sh**. Herefter bør I have 77 lande tilbage.

_Hint:_ Hvis du vil tjekke om en variabel indeholder manglende observationer kan du bruge en af de to komplementære kommandoer:
```py
df['sh'].notnull()
df['sh'].isnull()

```

**Din kode:**

In [4]:
df = pd.read_stata("PS3.dta")
df = df[df.grade != "D"]
df = df[df['sh'].notnull()]
df

country iso3       y03       y60        gy        sk        sh       u  \
1          USA  USA  1.000000  1.000000  0.017709  0.189912  0.115346  12.049   
2      Ireland  IRL  0.971402  0.423321  0.037025  0.209641  0.125219   9.351   
3       Norway  NOR  0.968075  0.761374  0.023295  0.283681  0.103698  11.848   
4      Belgium  BEL  0.906815  0.663650  0.024969  0.224658  0.101057   9.338   
5      Austria  AUT  0.880988  0.554981  0.028456  0.240480  0.095199   8.354   
..         ...  ...       ...       ...       ...       ...       ...     ...   
91    Tanzania  TZA  0.025963  0.028716  0.015365  0.048912  0.006915   2.705   
92      Malawi  MWI  0.023686  0.026768  0.014864  0.096660  0.011803   3.204   
93  Madagascar  MDG  0.023345  0.078392 -0.010462  0.039704  0.030596     NaN   
94    Ethiopia  ETH  0.023061  0.026356  0.014604  0.035875  0.014008   9.988   
95     Burundi  BDI  0.021141  0.037568  0.004339  0.038041  0.005237     NaN   

        isi         n grade  
1   0.97333  0.016080     A  
2   0.76672  0.010019     A  
3   0.87272  0.012205     A  
4   0.86572  0.004503     A  
5   0.86356  0.002440     A  
..      ...       ...   ...  
91  0.27567  0.028053     C  
92  0.25133  0.025739     C  
93  0.23800  0.025138     C  
94  0.19933  0.021697     C  
95  0.26390  0.016599     C  

[77 rows x 11 columns]

---
#### --- INTERMISSION --- 
Inden vi går videre med de næste opgaver, skal I lære en smart genvej i `statsmodels`, som gør det lidt simplere at køre regressioner. 

Indtil videre har vi manuelt udformet vores $X$-matrix, tilføjet en konstant og udformet vores $y$-vektor. Alt det kan vi faktisk springe over og i stedet bare specificere vores ønskede model med en tekststreng. Til det formål skal vi lige importere et nyt model fra statsmodels:

```py
import statsmodels.formula.api as smf
```

Prøv nu fx at regressere BNP vækst (`gy`) på befolkningstilvækst (`n`) og investeringsraten i kapital per indbygger (`sk`). Brug kommandoen
```py
results = smf.ols('gy ~ n + sk', data=df).fit()
print(results.summary())
```

Som du kan se af outputtet sørger statsmodels selv for at tilføje en konstant. Hvis du vil tilføje flere forklarende variable til regressionen, skal du bare tilføje flere led til tekststrengen


**Din kode:** 

In [5]:
import statsmodels.formula.api as smf
results = smf.ols('gy ~ n + sk', data=df).fit()
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                     gy   R-squared:                       0.277
Model:                            OLS   Adj. R-squared:                  0.258
Method:                 Least Squares   F-statistic:                     14.18
Date:                Fri, 27 Sep 2024   Prob (F-statistic):           6.11e-06
Time:                        08:34:14   Log-Likelihood:                 242.27
No. Observations:                  77   AIC:                            -478.5
Df Residuals:                      74   BIC:                            -471.5
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.0118      0.005      2.466      0.0

--- 

#### Opgave 2
Estimer den udvidede model (1) med `statsmodels`. Hvad er fortolkningen af $ \widehat{\beta}_1 $? Sammenlign estimatet af $ \beta_1 $ fra spørgsmål 7 i ugeseddel 3.



**Din kode:**

In [15]:
df['strucK'] = np.log(df.sk) - np.log(df.n + 0.075)
df['strucH'] = np.log(df.sh) - np.log(df.n + 0.075)
df['ln_y60'] = np.log(df.y60)

results1 = smf.ols('gy ~ ln_y60 + strucK + strucH', data=df).fit()
print(results1.summary())


                            OLS Regression Results                            
Dep. Variable:                     gy   R-squared:                       0.418
Model:                            OLS   Adj. R-squared:                  0.394
Method:                 Least Squares   F-statistic:                     17.50
Date:                Fri, 27 Sep 2024   Prob (F-statistic):           1.17e-08
Time:                        08:52:51   Log-Likelihood:                 250.63
No. Observations:                  77   AIC:                            -493.3
Df Residuals:                      73   BIC:                            -483.9
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.0029      0.003      0.998      0.3

**Dit svar:**

Når du går 1 ud af x stiger vil ygi falde med -0.0075. 

#### Opgave 3
Hvad er nul- og alternativ hypotesen for modellens samlede signifikans? Hvilket test skal I bruge for at teste nulhypotesen? Hvad er teststørrelsen og den kritiske værdi, og kan nulhypotesen afvises? Vær præcise, når I formulerer nul- og alternativ hypotesen i ord.

_Hint:_ for at finde den kritiske værdi til et _F_-test kan du bruge Python-pakken scipy:
```py
import scipy
q = ... # det ønskede signifikansniveau
dfn = ... # antal frihedsgrader i tælleren
dfd = ... # antal frihedsgrader i nævneren
scipy.stats.f.ppf(1-q, dfn, dfd)
```

**Din kode:**

In [7]:
import scipy
q = 0.05 # det ønskede signifikansniveau
dfn = 3 # antal frihedsgrader i tælleren
dfd = 73 # antal frihedsgrader i nævneren
scipy.stats.f.ppf(1-q, dfn, dfd)

2.7300187139961727

**Dit svar:**

F-test størrelse aflæser vi i forriga sumary til 17.5 hvilket er meget højere end vores kritiske så vi forkaster. I F-test tester den automatisk den samlede signigikans


#### Opgave 4
Hvad er nul- og alternativ hypoteserne for hver af de individuelle forklarende variable? Hvilket test skal I bruge for at teste nulhypoteserne? Hvad er teststørrelserne og den kritiske værdi, og kan nulhypoterne afvises? Vær præcise, når I formulerer nul- og alternativ hypoteserne i ord.


**Din kode:**

Vi kan igen kigge i vores forrige summarry og se at t-test for hvem af dem er signifikante, undtagen den sidste som er lige på grænsen. T-testene står der alle samemen.  

**Dit svar:**

In [13]:
#Kritisk t værdi
scipy.stats.t.ppf(0.975, 77)

#Kritisk normal værdi
scipy.stats.norm.ppf(0.975)

1.959963984540054

Ser igen at mine test-størrelser er større end kritiske værdier. Så vi forkaster igen. 


#### Opgave 5
Hvad er en passende alternativ hypotese, når I tester $ \beta_1 = 0 $ i model (1)? Udfør testet. 

[Hint: Hvilken betydning har hypotesen om betinget konvergens for $ \beta_1 $? Tag udgangspunkt i gruppespørgsmål i ugeseddel 3]


**Dit svar:**

Har testet foroven i t-test hvor $\beta_1=0$ der giver hylende lav p-værdi. Det vil sige at vi forkaster at det kun er de strukturelle der har indflydelse. Alternativ hypotese er at $\beta_1<0$. Ud fra testen ser vi at der er betinget konvergens, da $\beta_2\neq 0\neq \beta_3$, vi har også at $\beta_1\neq0$ som bare peger på den også har indflydelse


#### Opgave 6
I skal nu undersøge, om de strukturelle karakteristika samlet set kan udelades fra regressionsmodellen. Opskriv den relevante nulhypotese til F-testet ved at tage udgangspunkt i gruppespørgsmålene.

1. Udregn _F_-testet i hånden. I skal tage udgangspunkt i ligning [4.37] i Wooldridge (7. udgave). I skal bruge resultaterne fra spørgsmål 2 samt estimere en restrikteret model, hvor nulhypotesen fra gruppespørgsmålene er pålagt.

2. Lad Python udregne _F_-testet. Hvis dit OLS resultat-objekt fra statsmodels hedder `results` kan du bruge følgende kommando:

    ```py
    hypotheses = '(strucK = strucH = 0)'
    f_test = results.f_test(hypotheses)
    print(f_test)
    ```
3. Tjek at resultaterne  er ens.


**Din kode:**

In [35]:
# Manual calculation of F_test
## 1. Get the SSR of the restricted model
Y = df.gy
X = np.log(df.y60)

X = sm.add_constant(X)
model = sm.OLS(Y,X)
results = model.fit()
print(results.summary())
sum(results.resid**2)
#results2 = smf.ols(df['gy'],df['ln_y60']).fit()
#print(results2.summary())

## 2. Get the SSR of the unrestricted model
sum(results1.resid**2)
    
## 3. Calculate the F-test
tæller = (sum(results.resid**2)-sum(results1.resid**2))/2
nævner = (sum(results1.resid**2)/(77-3-1)) 
print(tæller/nævner)


## 4. Print critical value
scipy.stats.f.ppf(1-0.05, 2, 73)

                            OLS Regression Results                            
Dep. Variable:                     gy   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                 -0.013
Method:                 Least Squares   F-statistic:                   0.01383
Date:                Fri, 27 Sep 2024   Prob (F-statistic):              0.907
Time:                        09:25:46   Log-Likelihood:                 229.78
No. Observations:                  77   AIC:                            -455.6
Df Residuals:                      75   BIC:                            -450.9
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.0165      0.003      6.466      0.0

3.1221029319882803

In [40]:
# Statsmodels F-test
hypotheses = '(strucK = strucH = 0)'
f_test = results1.f_test(hypotheses)
print(f_test)

<F test: F=26.233555200023062, p=2.599819464276364e-09, df_denom=73, df_num=2>


**Dit svar:**


#### Opgave 7
I dette spørgsmål skal I implementere jeres egen version af OLS-estimatoren i MLR-tilfældet – ligesom i Problem Set 3. Denne gang skal jeres implementerede funktion dog både returnere OLS-parameterestimaterne og deres standardfejl.
 
Ligning (2) og (3) nedenfor viser formlen for OLS-estimatoren med homoskedastiske standardfejl.

$$
\widehat{\beta} = (X'X)^{-1}X'y \tag{2}
$$

$$
\text{var}(\widehat{\beta} | X) = \sigma^2 (X'X)^{-1}, \quad \sigma^2 = \frac{1}{N-K} \widehat{U}' \widehat{U} \tag{3}
$$

hvor 
- $ N $ er antallet af observationer, 
- $ K $ er antallet af parametre inkl. et konstantled 
- $ \widehat{U} \equiv y - X\widehat{\beta} $ er residualerne.
- $ \sigma^2 $ er residualvariansen
- Standardfejlene er kvadratroden af diagonalelementerne i kovarians-matricen $\text{var}(\widehat{\beta} | X)$

Brug din funktion til at estimere model (1) og bekræft, at du får de samme parameterestimator og standardfejl som statsmodels giver.

_Tip:_ Du kan tilgå diagonalelementerne af en matrix $A$ i numpy ved hjælp af `np.diagonal(A)`

In [68]:
y = df.gy
X = df[['ln_y60','strucK', 'strucH']]
X = sm.add_constant(X)
X = X.values

def ols(X, y):
    N = X.shape[0] # din kode her
    K = X.shape[1] # din kode her
    beta_hat = np.linalg.inv(X.T @ X) @ X.T @ Y # din kode her
    U_hat = y-X@beta_hat # din kode her
    sigma2 = 1/(N-K)*U_hat.T @ U_hat # din kode her
    var_beta_hat = sigma2*np.linalg.inv(X.T @ X) # din kode her
    se_ols = np.sqrt(np.diagonal(var_beta_hat)) # din kode her
    
    return beta_hat, se_ols

print(ols(X,y))

(array([ 0.0028857 , -0.00746327,  0.01237139,  0.00537258]), array([0.00289271, 0.00168056, 0.00268312, 0.0027139 ]))


Kør nedenstående celle for at estimere model (1) ved hjælp af din egen funktion:

In [70]:
X = df[['ln_y60', 'strucK', 'strucH']].values
X = sm.add_constant(X)
y = df['gy']

beta_hat, se_beta_hat = ols(X, y)

print('vars')
print('const, ln_y60, strucK, strucH')
print('beta_hat')
print(beta_hat.round(4))
print('se_beta_hat')
print(se_beta_hat.round(4))

vars
const, ln_y60, strucK, strucH
beta_hat
[ 0.0029 -0.0075  0.0124  0.0054]
se_beta_hat
[0.0029 0.0017 0.0027 0.0027]
